streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -364,22 +423,14 @@ public void restore() {
                     final ConsumerRecord<byte[], byte[]> record = iterator.next();
                     if (record.key() == null) {
                         log.warn("Read changelog record with null key from changelog {} at offset {}, " +
-                            "skipping it for restoration", changelogMetadata.changelogPartition, record.offset());
-                        iterator.remove();
+                            "skipping it for restoration", changelogMetadata.storeMetadata.changelogPartition(), record.offset());
                     } else {
+                        changelogMetadata.bufferedRecords.add(record);
                         final long offset = record.offset();
                         if (offset < limitOffset)
-                            changelogMetadata.bufferedLimitIndex++;
+                            changelogMetadata.bufferedLimitIndex = changelogMetadata.bufferedRecords.size();
                     }
                 }
-
-                // this is an optimization: if there's no buffered records so far, then we can reuse
```

Comment:
We have to give up this optimization since the records returned from consumer is `unmodifiable` view, so we cannot call `clear` on it directly.